gitlab/resource_gitlab_instance_cluster.go

```diff
@@ -0,0 +1,259 @@
+	client := meta.(*gitlab.Client)
+
+	clusterId, err := instanceClusterIdFromId(d.Id())
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] read gitlab instance cluster %d", clusterId)
+
+	cluster, _, err := client.InstanceCluster.GetCluster(clusterId)
+	if err != nil {
+		return err
+	}
+
+	d.Set("name", cluster.Name)
+	d.Set("domain", cluster.Domain)
+	d.Set("created_at", cluster.CreatedAt.String())
+	d.Set("provider_type", cluster.ProviderType)
+	d.Set("platform_type", cluster.PlatformType)
+	d.Set("environment_scope", cluster.EnvironmentScope)
+	d.Set("cluster_type", cluster.ClusterType)
+
+	d.Set("kubernetes_api_url", cluster.PlatformKubernetes.APIURL)
+	d.Set("kubernetes_ca_cert", cluster.PlatformKubernetes.CaCert)
+	d.Set("kubernetes_namespace", cluster.PlatformKubernetes.Namespace)
+	d.Set("kubernetes_authorization_type", cluster.PlatformKubernetes.AuthorizationType)
+
+	if cluster.ManagementProject == nil {
+		d.Set("management_project_id", "")
+	} else {
+		d.Set("management_project_id", strconv.Itoa(cluster.ManagementProject.ID))
+	}
+
+	return nil
+}
+
+func resourceGitlabInstanceClusterUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+
+	clusterId, err := instanceClusterIdFromId(d.Id())
+	if err != nil {
+		return err
+	}
+
+	options := &gitlab.EditClusterOptions{}
+
+	if d.HasChange("name") {
+		options.Name = gitlab.String(d.Get("name").(string))
+	}
+
+	if d.HasChange("domain") {
+		options.Domain = gitlab.String(d.Get("domain").(string))
+	}
+
+	if d.HasChange("environment_scope") {
+		options.EnvironmentScope = gitlab.String(d.Get("environment_scope").(string))
+	}
+
+	pk := &gitlab.EditPlatformKubernetesOptions{}
+
+	if d.HasChange("kubernetes_api_url") {
+		pk.APIURL = gitlab.String(d.Get("kubernetes_api_url").(string))
+	}
+
+	if d.HasChange("kubernetes_token") {
+		pk.Token = gitlab.String(d.Get("kubernetes_token").(string))
+	}
+
+	if d.HasChange("kubernetes_ca_cert") {
+		pk.CaCert = gitlab.String(d.Get("kubernetes_ca_cert").(string))
+	}
+
+	if *pk != (gitlab.EditPlatformKubernetesOptions{}) {
+		options.PlatformKubernetes = pk
+	}
+
+	if d.HasChange("management_project_id") {
+		options.ManagementProjectID = gitlab.String(d.Get("management_project_id").(string))
+	}
+
+	if *options != (gitlab.EditClusterOptions{}) {
+		log.Printf("[DEBUG] update gitlab instance cluster %d", clusterId)
+		_, _, err := client.InstanceCluster.EditCluster(clusterId, options)
+		if err != nil {
+			return err
+		}
+	}
+
+	return resourceGitlabInstanceClusterRead(d, meta)
+}
+
+func resourceGitlabInstanceClusterDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	clusterId, err := instanceClusterIdFromId(d.Id())
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] delete gitlab instance cluster %d", clusterId)
+
+	_, err = client.InstanceCluster.DeleteCluster(clusterId)
+
+	return err
+}
+
+func instanceClusterIdFromId(id string) (int, error) {
```

Comment:
IMO just using `strconv.Atoi(id)` inline would be clearer, plus there isn't a a situation where you would encounter a string conversion error on the resource ID other than during a resource import, so the error message here doesn't really add anything.